File: python/Infernux/engine/ui/toolbar_panel.py
"""Compact Notion-style toolbar above the scene viewport."""

from typing import Optional, TYPE_CHECKING
from Infernux.lib import InxGUIRenderable, InxGUIContext
from Infernux.engine.play_mode import PlayModeManager, PlayModeState
from Infernux.engine.i18n import t
from .editor_panel import EditorPanel
from .closable_panel import ClosablePanel
from .panel_registry import editor_panel
from .theme import Theme, ImGuiCol, ImGuiStyleVar


_CAMERA_DEFAULTS = {
    "fov": 60.0,
    "rotation_speed": 0.05,
    "pan_speed": 1.0,
    "zoom_speed": 1.0,
    "move_speed": 5.0,
    "move_speed_boost": 3.0,
}


def _noop():
    pass


@editor_panel("Toolbar", type_id="toolbar", title_key="panel.toolbar")
class ToolbarPanel(EditorPanel):
    WINDOW_TYPE_ID = "toolbar"
    WINDOW_DISPLAY_NAME = "Toolbar"

    def __init__(self, title: str = "Toolbar", engine=None,
                 play_mode_manager: Optional[PlayModeManager] = None):
        super().__init__(title, window_id="toolbar")
        self._engine = engine
        self._play_mode_manager = play_mode_manager
        self._camera_settings = dict(_CAMERA_DEFAULTS)
        if self._engine and self._play_mode_manager is None:
            self._play_mode_manager = self._engine.get_play_mode_manager()
        self._sync_camera_settings_from_engine()

    def set_engine(self, engine):
        self._engine = engine
        if engine:
            self._play_mode_manager = engine.get_play_mode_manager()
        self._apply_camera_settings()

    def set_play_mode_manager(self, manager: PlayModeManager):
        self._play_mode_manager = manager

    # ------------------------------------------------------------------
    # EditorPanel hooks
    # ------------------------------------------------------------------

    def _window_flags(self) -> int:
        return Theme.WINDOW_FLAGS_NO_SCROLL

    def _initial_size(self):
        return (800, 32)

    def _push_window_style(self, ctx):
        Theme.push_toolbar_vars(ctx)  # 5 style vars

    def _pop_window_style(self, ctx):
        ctx.pop_style_var(5)

    def on_render_content(self, ctx: InxGUIContext):
            win_w = ctx.get_window_width()
            state = self._get_state()
            is_playing = state in (PlayModeState.PLAYING, PlayModeState.PAUSED)
            is_paused  = state == PlayModeState.PAUSED

            # --- Right-aligned dropdowns (measure from right edge) ---
            right_x = win_w - 200.0
            if right_x < 300.0:
                right_x = 300.0

            # --- Centered play controls ---
            btn_w = 160.0
            cx = (win_w - btn_w) * 0.5
            if cx < 6.0:
                cx = 6.0
            ctx.set_cursor_pos_x(cx)

            # Play / Stop
            if is_playing and not is_paused:
                Theme.push_flat_button_style(ctx, *Theme.PLAY_ACTIVE)
            else:
                Theme.push_flat_button_style(ctx, *Theme.BTN_IDLE)
            ctx.button(t("toolbar.stop") if is_playing else t("toolbar.play"), self._on_play)
            ctx.pop_style_color(3)

            ctx.same_line(0, 2)

            # Pause / Resume
            if not is_playing:
                Theme.push_flat_button_style(ctx, *Theme.BTN_DISABLED)
            elif is_paused:
                Theme.push_flat_button_style(ctx, *Theme.PAUSE_ACTIVE)
            else:
                Theme.push_flat_button_style(ctx, *Theme.BTN_IDLE)
            ctx.button(t("toolbar.resume") if is_paused else t("toolbar.pause"),
                       self._on_pause if is_playing else _noop)
            ctx.pop_style_color(3)

            ctx.same_line(0, 2)

            # Step
            if is_paused:
                Theme.push_flat_button_style(ctx, *Theme.BTN_IDLE)
            else:
                Theme.push_flat_button_style(ctx, *Theme.BTN_DISABLED)
            ctx.button(t("toolbar.step"), self._on_step if is_paused else _noop)
            ctx.pop_style_color(3)

            # Time label while playing
            if is_playing:
                ctx.same_line(0, 8)
                tag = t("toolbar.status_paused") if is_paused else t("toolbar.status_playing")
                ctx.label(f"{tag}  {self._time_str()}")

            # --- Right dropdowns ---
            ctx.same_line(right_x)
            self._ghost_btn(ctx, t("toolbar.gizmos"), "##giz")
            if ctx.begin_popup("##giz"):
                Theme.push_popup_vars(ctx)  # 3 style vars
                self._popup_gizmos(ctx)
                ctx.pop_style_var(3)
                ctx.end_popup()

            ctx.same_line(0, 4)
            self._ghost_btn(ctx, t("toolbar.camera"), "##cam")
            if ctx.begin_popup("##cam"):
                Theme.push_popup_vars(ctx)  # 3 style vars
                self._popup_camera(ctx)
                ctx.pop_style_var(3)
                ctx.end_popup()


    # ---------- ghost button (Notion-style: transparent bg, subtle hover) ----
    @staticmethod
    def _ghost_btn(ctx: InxGUIContext, label: str, popup_id: str):
        Theme.push_ghost_button_style(ctx)
        ctx.button(label, lambda pid=popup_id: ctx.open_popup(pid))
        ctx.pop_style_color(3)

    # ---------- flat button helper ----
    @staticmethod
    def _flat_btn(ctx: InxGUIContext, base: tuple):
        Theme.push_flat_button_style(ctx, *base)

    # ---------- popup contents ----
    def _popup_gizmos(self, ctx: InxGUIContext):
        if not self._engine:
            ctx.label(t("toolbar.engine_not_available"))
            return
        ctx.dummy(200, 0)  # force minimum popup width
        ctx.label(t("toolbar.gizmos_header"))
        ctx.separator()
        ctx.dummy(0, 4)
        g = self._engine.is_show_grid()
        ng = ctx.checkbox(t("toolbar.show_grid"), g)
        if ng != g:
            self._engine.set_show_grid(ng)
        ctx.dummy(0, 4)

    # (key, label_key, min, max, step, step_fast, section_header_key)
    _CAMERA_PARAMS = (
        ("fov",              "toolbar.field_of_view",        10.0,  120.0, 1.0,   10.0, None),
        ("rotation_speed",   "toolbar.rotation_sensitivity", 0.005, 1.0,   0.005, 0.05, "toolbar.navigation_header"),
        ("pan_speed",        "toolbar.pan_speed",            0.1,   10.0,  0.1,   1.0,  None),
        ("zoom_speed",       "toolbar.zoom_speed",           0.1,   10.0,  0.1,   1.0,  None),
        ("move_speed",       "toolbar.move_speed",           0.1,   50.0,  0.1,   1.0,  None),
        ("move_speed_boost", "toolbar.speed_boost",          1.0,   20.0,  0.1,   1.0,  None),
    )

    def _popup_camera(self, ctx: InxGUIContext):
        cam = self._engine.editor_camera if self._engine else None
        if not cam:
            ctx.label(t("toolbar.camera_not_available"))
            return
        self._sync_camera_settings_from_engine()
        ctx.dummy(360, 0)  # force minimum popup width
        ctx.label(t("toolbar.scene_camera"))
        ctx.separator()
        ctx.dummy(0, 4)
        for key, label_key, mn, mx, step, step_fast, header in self._CAMERA_PARAMS:
            if header:
                ctx.label(t(header))
                ctx.separator()
                ctx.dummy(0, 4)
            self._set_camera_setting(
                key,
                self._render_camera_float_setting(
                    ctx, t(label_key), key,
                    self._camera_settings[key],
                    mn, mx, step=step, step_fast=step_fast,
                ),
            )
            ctx.dummy(0, 4)
        ctx.dummy(0, 2)
        ctx.button(t("toolbar.reset_camera_settings"), self._reset_camera_settings, width=-1.0)
        ctx.dummy(0, 4)

    def save_state(self) -> dict:
        self._sync_camera_settings_from_engine()
        return {
            "camera_settings": dict(self._camera_settings),
        }

    def load_state(self, data: dict) -> None:
        settings = data.get("camera_settings", {}) if isinstance(data, dict) else {}
        for key, default in _CAMERA_DEFAULTS.items():
            value = settings.get(key)
            if isinstance(value, (int, float)):
                self._camera_settings[key] = float(value)
            else:
                self._camera_settings[key] = float(default)
        self._apply_camera_settings()

    def _sync_camera_settings_from_engine(self):
        cam = self._engine.editor_camera if self._engine else None
        if not cam:
            return
        self._camera_settings.update({
            "fov": float(cam.fov),
            "rotation_speed": float(cam.rotation_speed),
            "pan_speed": float(cam.pan_speed),
            "zoom_speed": float(cam.zoom_speed),
            "move_speed": float(cam.move_speed),
            "move_speed_boost": float(cam.move_speed_boost),
        })

    def _apply_camera_settings(self):
        cam = self._engine.editor_camera if self._engine else None
        if not cam:
            return
        cam.fov = self._camera_settings["fov"]
        cam.rotation_speed = self._camera_settings["rotation_speed"]
        cam.pan_speed = self._camera_settings["pan_speed"]
        cam.zoom_speed = self._camera_settings["zoom_speed"]
        cam.move_speed = self._camera_settings["move_speed"]
        cam.move_speed_boost = self._camera_settings["move_speed_boost"]

    def _set_camera_setting(self, key: str, value: float, tolerance: float = 1e-4):
        clamped = float(value)
        if abs(clamped - self._camera_settings.get(key, clamped)) <= tolerance:
            return
        self._camera_settings[key] = clamped
        self._apply_camera_settings()

    def _reset_camera_settings(self):
        self._camera_settings = dict(_CAMERA_DEFAULTS)
        self._apply_camera_settings()

    @staticmethod
    def _render_camera_float_setting(
        ctx: InxGUIContext,
        label: str,
        key: str,
        value: float,
        min_value: float,
        max_value: float,
        *,
        step: float,
        step_fast: float,
    ) -> float:
        ctx.label(label)
        ctx.same_line(145)
        ctx.set_next_item_width(120)
        updated = ctx.float_slider(f"##{key}_slider", float(value), float(min_value), float(max_value))
        ctx.same_line(0, 6)
        ctx.set_next_item_width(72)
        updated = ctx.input_float(f"##{key}_input", float(updated), step, step_fast, 0)
        if updated < min_value:
            return float(min_value)
        if updated > max_value:
            return float(max_value)
        return float(updated)

    # ---------- state helpers ----
    def _get_state(self) -> PlayModeState:
        return self._play_mode_manager.state if self._play_mode_manager else PlayModeState.EDIT

    def _time_str(self) -> str:
        if not self._play_mode_manager:
            return "00:00.000"
        t = self._play_mode_manager.total_play_time
        return f"{int(t//60):02d}:{t%60:06.3f}"

    def _on_play(self):
        if not self._play_mode_manager:
            return
        if self._play_mode_manager.is_playing:
            self._play_mode_manager.exit_play_mode()
        else:
            if self._play_mode_manager.enter_play_mode():
                ClosablePanel.focus_panel_by_id("game_view")
                if self._engine:
                    self._engine.select_docked_window("game_view")

    def _on_pause(self):
        if not self._play_mode_manager:
            return
        self._play_mode_manager.toggle_pause()

    def _on_step(self):
        if not self._play_mode_manager:
            return
        self._play_mode_manager.step_frame()

File: cpp/infernux/function/editor/ToolbarPanel.cpp
#include "ToolbarPanel.h"

#include <algorithm>
#include <cstdio>

namespace infernux
{

// ════════════════════════════════════════════════════════════════════
// Construction
// ════════════════════════════════════════════════════════════════════

ToolbarPanel::ToolbarPanel() : EditorPanel("Toolbar", "toolbar")
{
}

// ════════════════════════════════════════════════════════════════════
// Camera settings
// ════════════════════════════════════════════════════════════════════

ToolbarPanel::CameraSettings ToolbarPanel::GetCameraSettings() const
{
    return m_cameraSettings;
}

void ToolbarPanel::SetCameraSettings(const CameraSettings &settings)
{
    m_cameraSettings = settings;
    if (applyCameraToEngine)
        applyCameraToEngine(m_cameraSettings);
}

// ════════════════════════════════════════════════════════════════════
// Translation helper
// ════════════════════════════════════════════════════════════════════

std::string ToolbarPanel::T(const std::string &key) const
{
    if (translate)
        return translate(key);
    // Fallback: return the key suffix after the last dot
    auto dot = key.rfind('.');
    return (dot != std::string::npos) ? key.substr(dot + 1) : key;
}

// ════════════════════════════════════════════════════════════════════
// Window flags & pre-render
// ════════════════════════════════════════════════════════════════════

ImGuiWindowFlags ToolbarPanel::GetWindowFlags() const
{
    return ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoScrollWithMouse;
}

void ToolbarPanel::PreRender(InxGUIContext * /*ctx*/)
{
    // Push toolbar spacing vars (5 vars)
    ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, EditorTheme::TOOLBAR_WIN_PAD);
    ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, EditorTheme::TOOLBAR_FRAME_PAD);
    ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, EditorTheme::TOOLBAR_ITEM_SPC);
    ImGui::PushStyleVar(ImGuiStyleVar_FrameRounding, EditorTheme::TOOLBAR_FRAME_RND);
    ImGui::PushStyleVar(ImGuiStyleVar_FrameBorderSize, EditorTheme::TOOLBAR_FRAME_BRD);
}

// ════════════════════════════════════════════════════════════════════
// Main content
// ════════════════════════════════════════════════════════════════════

void ToolbarPanel::OnRenderContent(InxGUIContext *ctx)
{
    float winW = ctx->GetWindowWidth();
    RenderPlayControls(ctx, winW);
    RenderRightDropdowns(ctx, winW);
}

void ToolbarPanel::PostRender(InxGUIContext * /*ctx*/)
{
    // Pop the 5 style vars pushed in PreRender (always, even when collapsed)
    ImGui::PopStyleVar(5);
}

// ════════════════════════════════════════════════════════════════════
// Play controls (centered)
// ════════════════════════════════════════════════════════════════════

void ToolbarPanel::RenderPlayControls(InxGUIContext *ctx, float winW)
{
    PlayState state = getPlayState ? getPlayState() : PlayState::Edit;
    bool isPlaying = (state == PlayState::Playing || state == PlayState::Paused);
    bool isPaused = (state == PlayState::Paused);

    float btnW = 160.0f;
    float cx = (winW - btnW) * 0.5f;
    if (cx < 6.0f)
        cx = 6.0f;
    ImGui::SetCursorPosX(cx);

    // ── Play / Stop ──────────────────────────────────────────────
    if (isPlaying && !isPaused)
        EditorTheme::PushFlatButtonStyle(EditorTheme::PLAY_ACTIVE);
    else
        EditorTheme::PushFlatButtonStyle(EditorTheme::BTN_IDLE);

    std::string playLabel = isPlaying ? T("toolbar.stop") : T("toolbar.play");
    if (ImGui::Button(playLabel.c_str())) {
        if (onPlay)
            onPlay();
    }
    ImGui::PopStyleColor(3);

    ImGui::SameLine(0.0f, 2.0f);

    // ── Pause / Resume ───────────────────────────────────────────
    if (!isPlaying)
        EditorTheme::PushFlatButtonStyle(EditorTheme::BTN_DISABLED);
    else if (isPaused)
        EditorTheme::PushFlatButtonStyle(EditorTheme::PAUSE_ACTIVE);
    else
        EditorTheme::PushFlatButtonStyle(EditorTheme::BTN_IDLE);

    std::string pauseLabel = isPaused ? T("toolbar.resume") : T("toolbar.pause");
    if (ImGui::Button(pauseLabel.c_str())) {
        if (isPlaying && onPause)
            onPause();
    }
    ImGui::PopStyleColor(3);

    ImGui::SameLine(0.0f, 2.0f);

    // ── Step ─────────────────────────────────────────────────────
    if (isPaused)
        EditorTheme::PushFlatButtonStyle(EditorTheme::BTN_IDLE);
    else
        EditorTheme::PushFlatButtonStyle(EditorTheme::BTN_DISABLED);

    std::string stepLabel = T("toolbar.step");
    if (ImGui::Button(stepLabel.c_str())) {
        if (isPaused && onStep)
            onStep();
    }
    ImGui::PopStyleColor(3);

    // ── Time label while playing ─────────────────────────────────
    if (isPlaying) {
        ImGui::SameLine(0.0f, 8.0f);
        std::string tag = isPaused ? T("toolbar.status_paused") : T("toolbar.status_playing");
        std::string timeStr = getPlayTimeStr ? getPlayTimeStr() : "00:00.000";
        ImGui::TextUnformatted((tag + "  " + timeStr).c_str());
    }
}

// ════════════════════════════════════════════════════════════════════
// Right-aligned dropdowns
// ════════════════════════════════════════════════════════════════════

void ToolbarPanel::RenderRightDropdowns(InxGUIContext *ctx, float winW)
{
    float rightX = winW - 200.0f;
    if (rightX < 300.0f)
        rightX = 300.0f;

    ImGui::SameLine(rightX);

    // Gizmos dropdown
    EditorTheme::PushGhostButtonStyle();
    std::string gizLabel = T("toolbar.gizmos");
    if (ImGui::Button(gizLabel.c_str()))
        ImGui::OpenPopup("##giz");
    ImGui::PopStyleColor(3);

    if (ImGui::BeginPopup("##giz")) {
        ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, EditorTheme::POPUP_WIN_PAD);
        ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, EditorTheme::POPUP_ITEM_SPC);
        ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, EditorTheme::POPUP_FRAME_PAD);
        PopupGizmos(ctx);
        ImGui::PopStyleVar(3);
        ImGui::EndPopup();
    }

    ImGui::SameLine(0.0f, 4.0f);

    // Camera dropdown
    EditorTheme::PushGhostButtonStyle();
    std::string camLabel = T("toolbar.camera");
    if (ImGui::Button(camLabel.c_str()))
        ImGui::OpenPopup("##cam");
    ImGui::PopStyleColor(3);

    if (ImGui::BeginPopup("##cam")) {
        ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, EditorTheme::POPUP_WIN_PAD);
        ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, EditorTheme::POPUP_ITEM_SPC);
        ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, EditorTheme::POPUP_FRAME_PAD);
        PopupCamera(ctx);
        ImGui::PopStyleVar(3);
        ImGui::EndPopup();
    }
}

// ════════════════════════════════════════════════════════════════════
// Popup: Gizmos
// ════════════════════════════════════════════════════════════════════

void ToolbarPanel::PopupGizmos(InxGUIContext *ctx)
{
    if (!isShowGrid) {
        ImGui::TextUnformatted(T("toolbar.engine_not_available").c_str());
        return;
    }

    ImGui::Dummy(ImVec2(200.0f, 0.0f)); // minimum popup width
    ImGui::TextUnformatted(T("toolbar.gizmos_header").c_str());
    ImGui::Separator();
    ImGui::Dummy(ImVec2(0.0f, 4.0f));

    bool grid = isShowGrid();
    if (ImGui::Checkbox(T("toolbar.show_grid").c_str(), &grid)) {
        if (setShowGrid)
            setShowGrid(grid);
    }
    ImGui::Dummy(ImVec2(0.0f, 4.0f));
}

// ════════════════════════════════════════════════════════════════════
// Popup: Camera
// ════════════════════════════════════════════════════════════════════

void ToolbarPanel::PopupCamera(InxGUIContext *ctx)
{
    // Sync from engine
    if (syncCameraFromEngine)
        m_cameraSettings = syncCameraFromEngine();

    ImGui::Dummy(ImVec2(360.0f, 0.0f)); // minimum popup width
    ImGui::TextUnformatted(T("toolbar.scene_camera").c_str());
    ImGui::Separator();
    ImGui::Dummy(ImVec2(0.0f, 4.0f));

    struct CamParam
    {
        const char *key;
        float *value;
        float mn, mx, step, stepFast;
        const char *headerKey; // null if no header
    };

    CamParam params[] = {
        {"toolbar.field_of_view", &m_cameraSettings.fov, 10.0f, 120.0f, 1.0f, 10.0f, nullptr},
        {"toolbar.rotation_sensitivity", &m_cameraSettings.rotationSpeed, 0.005f, 1.0f, 0.005f, 0.05f,
         "toolbar.navigation_header"},
        {"toolbar.pan_speed", &m_cameraSettings.panSpeed, 0.1f, 10.0f, 0.1f, 1.0f, nullptr},
        {"toolbar.zoom_speed", &m_cameraSettings.zoomSpeed, 0.1f, 10.0f, 0.1f, 1.0f, nullptr},
        {"toolbar.move_speed", &m_cameraSettings.moveSpeed, 0.1f, 50.0f, 0.1f, 1.0f, nullptr},
        {"toolbar.speed_boost", &m_cameraSettings.moveSpeedBoost, 1.0f, 20.0f, 0.1f, 1.0f, nullptr},
    };

    bool changed = false;

    for (auto &p : params) {
        if (p.headerKey) {
            ImGui::TextUnformatted(T(p.headerKey).c_str());
            ImGui::Separator();
            ImGui::Dummy(ImVec2(0.0f, 4.0f));
        }

        ImGui::TextUnformatted(T(p.key).c_str());
        ImGui::SameLine(145.0f);

        char sliderId[64];
        snprintf(sliderId, sizeof(sliderId), "##%s_slider", p.key);
        ImGui::SetNextItemWidth(120.0f);
        float prev = *p.value;
        ImGui::SliderFloat(sliderId, p.value, p.mn, p.mx);

        ImGui::SameLine(0.0f, 6.0f);

        char inputId[64];
        snprintf(inputId, sizeof(inputId), "##%s_input", p.key);
        ImGui::SetNextItemWidth(72.0f);
        ImGui::InputFloat(inputId, p.value, p.step, p.stepFast, "%.3f");

        // Clamp
        *p.value = (std::min)((std::max)(*p.value, p.mn), p.mx);

        if (*p.value != prev)
            changed = true;

        ImGui::Dummy(ImVec2(0.0f, 4.0f));
    }

    // Reset button
    ImGui::Dummy(ImVec2(0.0f, 2.0f));
    if (ImGui::Button(T("toolbar.reset_camera_settings").c_str(), ImVec2(-1.0f, 0.0f))) {
        m_cameraSettings.fov = CAMERA_DEFAULTS_FOV;
        m_cameraSettings.rotationSpeed = CAMERA_DEFAULTS_ROTATION;
        m_cameraSettings.panSpeed = CAMERA_DEFAULTS_PAN;
        m_cameraSettings.zoomSpeed = CAMERA_DEFAULTS_ZOOM;
        m_cameraSettings.moveSpeed = CAMERA_DEFAULTS_MOVE;
        m_cameraSettings.moveSpeedBoost = CAMERA_DEFAULTS_BOOST;
        changed = true;
    }
    ImGui::Dummy(ImVec2(0.0f, 4.0f));

    if (changed && applyCameraToEngine)
        applyCameraToEngine(m_cameraSettings);
}

} // namespace infernux
